File: packages/neptune_flutter_ui/lib/src/widgets/neptune_secure_inputs.dart
// © 2026 Neptune.Fintech (neptune.ly) · Neptune Odyssey Community License v1.0

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import '../theme/extensions.dart';
import '../theme/neptune_theme.dart';

/// A row of boxed single-digit OTP cells (web `<npt-otp-input>`): auto-advance,
/// backspace-rewind, and paste-fill. Focused/filled cells lift to a
/// [ColorScheme.primary] border. Numeric only. Theme-only, RTL-safe (the cell
/// row stays LTR like the web `direction: ltr`). Emits [onChanged] on every
/// edit and [onCompleted] once all [length] cells are filled.
class NeptuneOtpInput extends StatefulWidget {
  final int length;
  final ValueChanged<String>? onChanged;
  final ValueChanged<String>? onCompleted;

  /// When true the digits render as dots (used by [NeptunePinInput]).
  final bool obscure;

  const NeptuneOtpInput({
    super.key,
    this.length = 6,
    this.onChanged,
    this.onCompleted,
    this.obscure = false,
  });

  @override
  State<NeptuneOtpInput> createState() => _NeptuneOtpInputState();
}

class _NeptuneOtpInputState extends State<NeptuneOtpInput> {
  late List<TextEditingController> _controllers;
  late List<FocusNode> _nodes;

  @override
  void initState() {
    super.initState();
    _build();
  }

  void _build() {
    _controllers = List.generate(widget.length, (_) => TextEditingController());
    _nodes = List.generate(widget.length, (_) => FocusNode());
  }

  @override
  void didUpdateWidget(covariant NeptuneOtpInput oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.length != widget.length) {
      _disposeAll();
      _build();
    }
  }

  void _disposeAll() {
    for (final c in _controllers) {
      c.dispose();
    }
    for (final n in _nodes) {
      n.dispose();
    }
  }

  @override
  void dispose() {
    _disposeAll();
    super.dispose();
  }

  String get _value => _controllers.map((c) => c.text).join();

  void _emit() {
    final value = _value;
    widget.onChanged?.call(value);
    if (value.length == widget.length) {
      widget.onCompleted?.call(value);
    }
  }

  void _onChanged(int index, String raw) {
    final digits = raw.replaceAll(RegExp(r'\D'), '');

    // Paste-fill: spread multiple digits across this cell and the following.
    if (digits.length > 1) {
      var i = index;
      for (final ch in digits.split('')) {
        if (i >= widget.length) break;
        _controllers[i].text = ch;
        i++;
      }
      final next = i < widget.length ? i : widget.length - 1;
      _nodes[next].requestFocus();
      _controllers[next].selection = TextSelection.collapsed(
        offset: _controllers[next].text.length,
      );
      setState(_emit);
      return;
    }

    final ch = digits.isEmpty ? '' : digits.characters.last;
    _controllers[index].text = ch;
    _controllers[index].selection =
        TextSelection.collapsed(offset: ch.length);
    if (ch.isNotEmpty && index < widget.length - 1) {
      _nodes[index + 1].requestFocus();
    }
    setState(_emit);
  }

  KeyEventResult _onKey(int index, FocusNode node, KeyEvent event) {
    if (event is! KeyDownEvent && event is! KeyRepeatEvent) {
      return KeyEventResult.ignored;
    }
    final key = event.logicalKey;
    if (key == LogicalKeyboardKey.backspace &&
        _controllers[index].text.isEmpty &&
        index > 0) {
      _controllers[index - 1].text = '';
      _nodes[index - 1].requestFocus();
      setState(_emit);
      return KeyEventResult.handled;
    }
    if (key == LogicalKeyboardKey.arrowLeft && index > 0) {
      _nodes[index - 1].requestFocus();
      return KeyEventResult.handled;
    }
    if (key == LogicalKeyboardKey.arrowRight && index < widget.length - 1) {
      _nodes[index + 1].requestFocus();
      return KeyEventResult.handled;
    }
    return KeyEventResult.ignored;
  }

  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: TextDirection.ltr,
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          for (var i = 0; i < widget.length; i++) ...[
            if (i > 0) const SizedBox(width: 8),
            _OtpCell(
              controller: _controllers[i],
              focusNode: _nodes[i],
              obscure: widget.obscure,
              isLast: i == widget.length - 1,
              onChanged: (raw) => _onChanged(i, raw),
              onKey: (node, event) => _onKey(i, node, event),
            ),
          ],
        ],
      ),
    );
  }
}

/// A single OTP/PIN cell. Filled or focused cells take the primary border.
class _OtpCell extends StatefulWidget {
  final TextEditingController controller;
  final FocusNode focusNode;
  final bool obscure;
  final bool isLast;
  final ValueChanged<String> onChanged;
  final KeyEventResult Function(FocusNode, KeyEvent) onKey;

  const _OtpCell({
    required this.controller,
    required this.focusNode,
    required this.obscure,
    required this.isLast,
    required this.onChanged,
    required this.onKey,
  });

  @override
  State<_OtpCell> createState() => _OtpCellState();
}

class _OtpCellState extends State<_OtpCell> {
  @override
  void initState() {
    super.initState();
    widget.focusNode.addListener(_onFocusChange);
    widget.controller.addListener(_onTextChange);
  }

  @override
  void dispose() {
    widget.focusNode.removeListener(_onFocusChange);
    widget.controller.removeListener(_onTextChange);
    super.dispose();
  }

  void _onFocusChange() {
    if (widget.focusNode.hasFocus) {
      widget.controller.selection = TextSelection(
        baseOffset: 0,
        extentOffset: widget.controller.text.length,
      );
    }
    setState(() {});
  }

  void _onTextChange() => setState(() {});

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final shape = Theme.of(context).extension<NptShape>()!;
    final type = Theme.of(context).extension<NptType>()!;
    final text = Theme.of(context).textTheme;

    final focused = widget.focusNode.hasFocus;
    final filled = widget.controller.text.isNotEmpty;
    final active = focused || filled;

    final cellStyle = (text.headlineSmall ?? text.titleLarge)?.copyWith(
      fontFamily: type.num,
      color: scheme.onSurface,
      fontFeatures: const [FontFeature.tabularFigures()],
    );

    return SizedBox(
      width: 48,
      height: 56,
      child: Focus(
        onKeyEvent: widget.onKey,
        child: TextField(
          controller: widget.controller,
          focusNode: widget.focusNode,
          keyboardType: TextInputType.number,
          textAlign: TextAlign.center,
          obscureText: widget.obscure,
          obscuringCharacter: '•',
          maxLength: 1,
          showCursor: true,
          style: cellStyle,
          textInputAction:
              widget.isLast ? TextInputAction.done : TextInputAction.next,
          onChanged: widget.onChanged,
          inputFormatters: [FilteringTextInputFormatter.digitsOnly],
          decoration: InputDecoration(
            counterText: '',
            isDense: true,
            filled: true,
            fillColor: scheme.surfaceContainerLowest,
            contentPadding: EdgeInsets.zero,
            enabledBorder: OutlineInputBorder(
              borderRadius: shape.rSm,
              borderSide: BorderSide(
                color: active ? scheme.primary : scheme.outline,
              ),
            ),
            focusedBorder: OutlineInputBorder(
              borderRadius: shape.rSm,
              borderSide: BorderSide(color: scheme.primary, width: 2),
            ),
          ),
        ),
      ),
    );
  }
}

/// A masked variant of [NeptuneOtpInput] (web `<npt-pin-input>`): the same boxed
/// cells, but always obscured with dots and defaulting to 4 digits.
class NeptunePinInput extends StatelessWidget {
  final int length;
  final ValueChanged<String>? onChanged;
  final ValueChanged<String>? onCompleted;

  const NeptunePinInput({
    super.key,
    this.length = 4,
    this.onChanged,
    this.onCompleted,
  });

  @override
  Widget build(BuildContext context) {
    return NeptuneOtpInput(
      length: length,
      obscure: true,
      onChanged: onChanged,
      onCompleted: onCompleted,
    );
  }
}

/// A 3×4 numeric keypad (web `<npt-amount-keypad>`): digits 1–9, a decimal '.',
/// 0, and a backspace. Each key is a large (>= 56dp) themed tile in the number
/// font. Emits the pressed key string ('0'–'9' or '.') via [onKey], and
/// [onBackspace] for the delete key. Theme-only, RTL-safe (logical grid).
class NeptuneAmountKeypad extends StatelessWidget {
  final ValueChanged<String>? onKey;
  final VoidCallback? onBackspace;

  const NeptuneAmountKeypad({super.key, this.onKey, this.onBackspace});

  @override
  Widget build(BuildContext context) {
    const keys = <String>[
      '1', '2', '3', //
      '4', '5', '6', //
      '7', '8', '9', //
      '.', '0', 'back', //
    ];

    return GridView.count(
      crossAxisCount: 3,
      shrinkWrap: true,
      physics: const NeverScrollableScrollPhysics(),
      mainAxisSpacing: 8,
      crossAxisSpacing: 8,
      childAspectRatio: 1.6,
      children: [
        for (final k in keys)
          _KeypadKey(
            value: k,
            onTap: () {
              if (k == 'back') {
                onBackspace?.call();
              } else {
                onKey?.call(k);
              }
            },
          ),
      ],
    );
  }
}

/// One keypad tile. Digits/decimal use the number font; backspace is an icon.
class _KeypadKey extends StatelessWidget {
  final String value;
  final VoidCallback onTap;

  const _KeypadKey({required this.value, required this.onTap});

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final shape = Theme.of(context).extension<NptShape>()!;
    final text = Theme.of(context).textTheme;
    final isAction = value == 'back' || value == '.';

    final keyStyle = NeptuneTheme.moneyStyle(
      context,
      base: text.headlineSmall ?? text.titleLarge,
    ).copyWith(color: scheme.onSurface);

    final Widget child = value == 'back'
        ? Icon(Icons.backspace_outlined,
            size: 24, color: scheme.onSurfaceVariant)
        : Text(value, style: keyStyle);

    return Material(
      color: isAction ? null : scheme.surfaceContainerHigh,
      borderRadius: shape.rMd,
      clipBehavior: Clip.antiAlias,
      child: InkWell(
        onTap: onTap,
        borderRadius: shape.rMd,
        child: Container(
          constraints: const BoxConstraints(minHeight: 56),
          alignment: Alignment.center,
          child: child,
        ),
      ),
    );
  }
}

File: packages/neptune_kmp_ui/odyssey-compose-ui/src/commonMain/kotlin/ly/neptune/odyssey/ui/components/NeptuneSecureInputs.kt
// © 2026 Neptune.Fintech (neptune.ly) · Neptune Odyssey Community License v1.0
//
// Secure-entry inputs. Web counterparts: `<npt-pin-input>`,
// `<npt-amount-keypad>` (money-inputs.ts) · Flutter:
// neptune_secure_inputs.dart. The PIN input is the boxed OTP entry, always
// dot-masked and defaulting to 4 cells; the amount keypad is a 3-column grid
// of ≥56dp keys (digits on surface-container-high, brand `md` corner, money
// type; decimal and backspace as quiet action keys) with a haptic tap per
// key. Theme-only, RTL-safe (the grid lays out logically and mirrors).

package ly.neptune.odyssey.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.hapticfeedback.HapticFeedback
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.max
import ly.neptune.odyssey.ui.glyphs.NptInputGlyphs
import ly.neptune.odyssey.ui.theme.NeptuneTheme
import ly.neptune.odyssey.ui.theme.NptFeedbackCue

/**
 * A masked PIN entry: the same boxed cells as [NeptuneOtpInput], but always
 * obscured with dots and defaulting to 4 digits.
 *
 * Web counterpart: `<npt-pin-input>` · Flutter: `NeptunePinInput`.
 *
 * [value] is digits-only, truncated to [length]; [onComplete] fires once
 * every cell is filled. [error] recolours the cell borders; the cell row
 * stays LTR under RTL (like the web `direction: ltr`).
 */
@Composable
public fun NeptunePinInput(
    value: String,
    onValueChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    length: Int = 4,
    error: Boolean = false,
    enabled: Boolean = true,
    onComplete: ((String) -> Unit)? = null,
) {
    NeptuneOtpInput(
        value = value,
        onValueChange = onValueChange,
        modifier = modifier,
        length = length,
        obscure = true,
        error = error,
        enabled = enabled,
        onComplete = onComplete,
    )
}

// The keypad rows: digits 1–9, then decimal · 0 · backspace.
private val keypadRows: List<List<String>> = listOf(
    listOf("1", "2", "3"),
    listOf("4", "5", "6"),
    listOf("7", "8", "9"),
    listOf(".", "0", "back"),
)

/**
 * A 3×4 on-screen amount keypad: digits 1–9, a decimal point, 0 and a
 * backspace. Digit keys are `surfaceContainerHigh` tiles on the brand `md`
 * corner in the money style (num family, tabular, headlineSmall); the
 * decimal and backspace render as quiet (transparent) action keys. Every
 * key is at least 56dp tall (8dp gutters, 1.6 aspect cells) and fires the
 * brand haptic tap.
 *
 * Web counterpart: `<npt-amount-keypad>` · Flutter: `NeptuneAmountKeypad`.
 *
 * [onKey] receives `"0"`–`"9"` or `"."`; [onBackspace] fires for the delete
 * key. Null callbacks disable their keys. The grid lays out logically, so
 * it mirrors under RTL like the web/Flutter grids.
 */
@Composable
public fun NeptuneAmountKeypad(
    onKey: ((String) -> Unit)?,
    onBackspace: (() -> Unit)?,
    modifier: Modifier = Modifier,
) {
    BoxWithConstraints(modifier) {
        // The Flutter grid: 3 columns, 8dp gutters, 1.6 aspect cells — with
        // the 56dp key floor. Unbounded slots fall back to shrink-wrapped
        // 90dp keys (the Segmented port's bounded-width guard).
        val bounded = maxWidth != Dp.Infinity
        val keyWidth = if (bounded) (maxWidth - 16.dp) / 3 else 90.dp
        val keyHeight = max(56.dp, keyWidth / 1.6f)
        Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
            keypadRows.forEach { row ->
                Row(
                    modifier = if (bounded) Modifier.fillMaxWidth() else Modifier,
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    row.forEach { key ->
                        KeypadKey(
                            value = key,
                            onTap = when {
                                key == "back" -> onBackspace
                                onKey != null -> ({ onKey(key) })
                                else -> null
                            },
                            modifier = (
                                if (bounded) {
                                    Modifier.weight(1f)
                                } else {
                                    Modifier.width(keyWidth)
                                }
                                ).height(keyHeight),
                        )
                    }
                }
            }
        }
    }
}

/** One keypad tile. Digits use the money style; backspace is a 24dp glyph. */
@Composable
private fun KeypadKey(
    value: String,
    onTap: (() -> Unit)?,
    modifier: Modifier = Modifier,
) {
    val scheme = MaterialTheme.colorScheme
    val typography = MaterialTheme.typography
    val shape = NeptuneTheme.shape.rMd
    val feedback = NeptuneTheme.feedback
    val haptics: HapticFeedback = LocalHapticFeedback.current
    val isBack = value == "back"
    // The decimal and backspace are quiet action keys (no tile fill).
    val isAction = isBack || value == "."
    val keyStyle = NeptuneTheme.moneyStyle(base = typography.headlineSmall)
        .copy(color = scheme.onSurface)

    Box(
        modifier = modifier
            .clip(shape)
            .then(if (isAction) Modifier else Modifier.background(scheme.surfaceContainerHigh))
            .clickable(enabled = onTap != null, role = Role.Button) {
                feedback.trigger(NptFeedbackCue.Tap, haptics)
                onTap?.invoke()
            },
        contentAlignment = Alignment.Center,
    ) {
        if (isBack) {
            Icon(
                NptInputGlyphs.backspace,
                contentDescription = "Backspace",
                tint = scheme.onSurfaceVariant,
                modifier = Modifier.size(24.dp),
            )
        } else {
            Text(text = NeptuneTheme.formatDigits(value), style = keyStyle)
        }
    }
}
